python/descriptor.c

```diff
@@ -38,113 +42,919 @@
+                                                 const char* msg_name) {
+  if (!self->options) {
+    // Load descriptors protos if they are not loaded already. We have to do
+    // this lazily, otherwise, it would lead to circular imports.
+    PyObject* mod = PyImport_ImportModule("google.protobuf.descriptor_pb2");
+    Py_DECREF(mod);
+
+    // Find the correct options message.
+    PyObject* default_pool = PyUpb_DescriptorPool_GetDefaultPool();
+    const upb_symtab* symtab = PyUpb_DescriptorPool_GetSymtab(default_pool);
+    const upb_msgdef* m = upb_symtab_lookupmsg(symtab, msg_name);
+    assert(m);
+
+    // Copy the options message from C to Python using serialize+parse.
+    // We don't wrap the C object directly because there is no guarantee that
+    // the descriptor_pb2 that was loaded at runtime has the same members or
+    // layout as the C types that were compiled in.
+    size_t size;
+    PyObject* py_arena = PyUpb_Arena_New();
+    upb_arena* arena = PyUpb_Arena_Get(py_arena);
+    char* pb = upb_encode(opts, layout, arena, &size);
+    upb_msg* opts2 = upb_msg_new(m, arena);
+    assert(opts2);
+    bool ok = _upb_decode(pb, size, opts2, upb_msgdef_layout(m),
+                          upb_symtab_extreg(symtab), 0,
+                          arena) == kUpb_DecodeStatus_Ok;
+    (void)ok;
+    assert(ok);
+
+    // Disabled until python/message.c is reviewed and checked in.
+    // self->options = PyUpb_CMessage_Get(opts2, m, py_arena);
+    Py_DECREF(py_arena);
+    PyErr_Format(PyExc_NotImplementedError, "Not yet implemented");
+    return NULL;
+  }
+
+  Py_INCREF(self->options);
+  return self->options;
+}
+
+typedef void* PyUpb_ToProto_Func(const void* def, upb_arena* arena);
+
+static PyObject* PyUpb_DescriptorBase_CopyToProto(PyObject* _self,
+                                                  PyUpb_ToProto_Func* func,
+                                                  const upb_msglayout* layout,
+                                                  PyObject* py_proto) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  upb_arena* arena = upb_arena_new();
+  upb_msg* proto = func(self->def, arena);
+  size_t size;
+  char* pb = upb_encode(proto, layout, arena, &size);
+  PyObject* str = PyBytes_FromStringAndSize(pb, size);
+  // Disabled until python/message.c is reviewed and checked in.
+  // PyObject *ret = PyUpb_CMessage_MergeFromString(py_proto, str);
+  upb_arena_free(arena);
+  Py_DECREF(str);
+  return PyErr_Format(PyExc_NotImplementedError, "Not yet implemented");
+}
+
+static void PyUpb_DescriptorBase_Dealloc(PyUpb_DescriptorBase* base) {
   PyUpb_ObjCache_Delete(base->def);
   Py_DECREF(base->pool);
-  PyObject_Del(self);
+  Py_XDECREF(base->options);
+  PyUpb_Dealloc(base);
+}
+
+#define DESCRIPTOR_BASE_SLOTS                           \
+  {Py_tp_new, (void*)&PyUpb_Forbidden_New}, {           \
+    Py_tp_dealloc, (void*)&PyUpb_DescriptorBase_Dealloc \
+  }
```

Comment:
the formatting here feel a bit offf.  did clang-format really do this?  can you force a new line between the two elements?